Objects/dictobject.c

```diff
@@ -1465,15 +1782,16 @@ dictresize(PyInterpreterState *interp, PyDictObject *mp,
         unicode = 0;
     }
 
+    ensure_shared_on_resize(mp);
     /* NOTE: Current odict checks mp->ma_keys to detect resize happen.
      * So we can't reuse oldkeys even if oldkeys->dk_size == newsize.
      * TODO: Try reusing oldkeys when reimplement odict.
      */
 
     /* Allocate a new table. */
-    mp->ma_keys = new_keys_object(interp, log2_newsize, unicode);
+    set_keys(mp, new_keys_object(interp, log2_newsize, unicode));
     if (mp->ma_keys == NULL) {
-        mp->ma_keys = oldkeys;
+        set_keys(mp, oldkeys);
```

Comment:
I don't think the possible temporary assignment to NULL is thread-safe without the GIL. Let's assign the result of `new_keys_object()` to a temporary variable and check that before assigning it to `mp->ma_keys`.

Reply:
When I was looking at this function when you made a comment about it on #114741 I realized that the assignment of it to a new empty table isn't right at all!  We really need to allocate the new table, copy everything over, and then we can publish it.